woocommerce/woocommerce-gutenberg-products-block
# Move `TextInput` to checkout package and allow it to be used for input type=number

As part of the Points and Rewards integration, I found the need to expose several of our UI components. This PR will make `TextInput` available via the `@woocommerce/blocks-checkout` package, and also allow it to be used for number inputs.

To allow additional props to be passed to the `<input>` element, I added a new prop to `ValidatedTextInput` and `TextInput` called `inputProps`, which will be spread onto the `<input>` element in `TextInput`.

Because of accessing the validation context from a different package, the validation functions returned by `useValidationContext` were not available, so I modified the `ValidatedTextInput` component to accept the functions as props, and also the `ValidatedTextInput` component.

### How to test the changes in this Pull Request:

1. Smoke test the cart/checkout flow.
2. Try adding valid _and_ invalid coupons in both the Cart and Checkout blocks.
3. Ensure validation still works in the Checkout form, invalid emails, empty fields etc.
4. Modify `assets/js/base/components/cart-checkout/address-form/address-form.js` and add the following at the top of the file and toward the top of the `AddressForm` function respectively:
 ```typescript
import { useState } from '@wordpress/element';
```
```typescript
const [ testValue, setTestValue ] = useState( 0 );
```
then output the following just before the closing `</div>` of the returned component:
```typescript
<ValidatedTextInput
  label={ 'Test input' }
  type={ 'number' }
  { ...textInputValidationFunctions }
  onChange={ setTestValue }
  value={ testValue }
  step={ 2 }
  min={ 0 }
  max={ 10 }
/>
``` 
5. The Address form should have an input of type `number` now, and the extra props you put in `inputProps` should be attributes on the element.


### Changelog

> Move `ValidatedTextInput` to the `@woocommerce/blocks-checkout` package.
